src/stash/src/lib.rs

```diff
@@ -0,0 +1,465 @@
+    /// The provided `upper` must be greater than or equal to the current upper
+    /// frontier.
+    ///
+    /// Intuitively, this method declares that all times less than `upper` are
+    /// definite.
+    pub fn seal(&self, upper: i64) -> Result<(), StashError> {
+        let mut conn = self.conn.lock().expect("lock poisoned");
+        let tx = conn.transaction()?;
+        assert!(self.upper_tx(&tx)? <= upper);
+        tx.execute(
+            "UPDATE uppers SET upper = $upper WHERE arrangement_id = $arrangement_id",
+            named_params! {"$upper": upper, "$arrangement_id": self.arrangement_id},
+        )?;
+        tx.commit()?;
+        Ok(())
+    }
+
+    /// Advances the since frontier to the specified value.
+    ///
+    /// The provided `since` must be greater than or equal to the current since
+    /// frontier but less than or equal to the current upper frontier.
+    ///
+    /// Intuitively, this method performs logical compaction. Existing entries
+    /// whose time is less than `since` are fast-forwarded to `since`.
+    pub fn compact(&self, since: i64) -> Result<(), StashError> {
+        let mut conn = self.conn.lock().expect("lock poisoned");
+        let tx = conn.transaction()?;
+        assert!(since <= self.upper_tx(&tx)?);
+        assert!(self.since_tx(&tx)? <= since);
+        tx.execute(
+            "UPDATE sinces SET since = $since WHERE arrangement_id = $arrangement_id",
+            named_params! {"$since": since, "$arrangement_id": self.arrangement_id},
+        )?;
+        tx.commit()?;
+        Ok(())
+    }
+
+    /// Consolidates entries less than the since frontier.
+    ///
+    /// Intuitively, this method performs physical compaction. Existing
+    /// key–value pairs whose time is less than the since frontier are
+    /// consolidated together when possible.
+    pub fn consolidate(&mut self) -> Result<(), StashError> {
+        let mut conn = self.conn.lock().expect("lock poisoned");
+        let tx = conn.transaction()?;
+        let since = self.since_tx(&tx)?;
+        tx.execute(
+            "INSERT INTO data (arrangement_id, key, value, time, diff)
+             SELECT arrangement_id, key, value, $since, sum(diff) FROM data
+             WHERE arrangement_id = $arrangement_id AND time < $since
+             GROUP BY key, value
+             ON CONFLICT (arrangement_id, key, value, time) DO UPDATE SET diff = diff + excluded.diff",
+            named_params! {
+                "$arrangement_id": self.arrangement_id,
+                "$since": since,
+            },
+        )?;
+        tx.execute(
+            "DELETE FROM data WHERE arrangement_id = $arrangement_id AND time < $since",
+            named_params! {
+                "$arrangement_id": self.arrangement_id,
+                "$since": since,
+            },
+        )?;
+        tx.commit()?;
+        Ok(())
+    }
+
+    /// Reports the current since frontier.
+    pub fn since(&self) -> Result<i64, StashError> {
```

Comment:
Since has a similar comment about frontiers as `upper`, that "being empty" is interesting. Setting `since` to the empty frontier is a way of communicating "we are done with this collection" (all future queries will be at times greater or equal to ... ah, no time then).

I don't think we use this at the moment, and the potential uses are maybe corner cases. But if it is easy to make into an `Option<i64>` or `Antichain<i64>` then great!